binding/Binding/HandleDictionary.cs

```diff
@@ -40,59 +35,37 @@ internal static bool GetInstance<TSkiaObject> (IntPtr handle, out TSkiaObject in
 			}
 		}
 
-		/// <summary>
-		/// Retrieve or create an instance for the native handle.
-		/// </summary>
-		/// <returns>The instance, or null if the handle was null.</returns>
-		internal static TSkiaObject GetObject<TSkiaObject, TSkiaImplementation> (IntPtr handle, bool owns = true, bool unrefExisting = true, bool refNew = false)
+		internal static bool TryGetObject<TSkiaObject> (IntPtr handle, bool owns, bool unrefExisting, bool refNew, out TSkiaObject obj)
 			where TSkiaObject : SKObject
-			where TSkiaImplementation : SKObject, TSkiaObject
 		{
-			if (handle == IntPtr.Zero)
-				return null;
+			if (handle == IntPtr.Zero) {
+				obj = null;
+				return true;
+			}
 
 			instancesLock.EnterUpgradeableReadLock ();
+
 			try {
-				if (GetInstanceNoLocks<TSkiaObject> (handle, out var instance)) {
+				if (GetInstanceNoLocks<TSkiaObject> (handle, out obj)) {
 					// some object get automatically referenced on the native side,
 					// but managed code just has the same reference
-					if (unrefExisting && instance is ISKReferenceCounted refcnt) {
+					if (unrefExisting && obj is ISKReferenceCounted refcnt) {
 #if THROW_OBJECT_EXCEPTIONS
 						if (refcnt.GetReferenceCount () == 1)
 							throw new InvalidOperationException (
 								$"About to unreference an object that has no references. " +
-								$"H: {handle.ToString ("x")} Type: {instance.GetType ()}");
+								$"H: {handle.ToString ("x")} Type: {obj.GetType ()}");
 #endif
 						refcnt.SafeUnRef ();
 					}
 
-					return instance;
+					return true;
 				}
-
-				var type = typeof (TSkiaImplementation);
-				var constructor = constructors.GetOrAdd (type, t => GetConstructor (t));
-
-				// we don't need to go into a writable here as the object will do it in the Handle property
-				var obj = (TSkiaObject)constructor.Invoke (new object[] { handle, owns });
```

Comment:
Just noticed the construction is no longer inside the `instancesLock`. This might allow for concurrent construction of the same object. We might need that delegate 😢

Reply:
Hmm

Reply:
`internal static TSkiaObject GetOrAddObject<TSkiaObject> (IntPtr handle, bool owns, bool unrefExisting, bool refNew, Func<IntPtr, bool, TSkiaObject> objectFactory)`

Reply:
Okay will adjust the implementation a bit

Reply:
`internal static GRContext GetObject (IntPtr handle) => GetOrAddObject (handle, (h,o) => new GRContext (h, o));`

Reply:
Yes, that looks like it should work now. And the delegate is reused instead of for each member.